docs/architecture/adr-020-protobuf-client-encoding.md

```diff
@@ -0,0 +1,105 @@
+# ADR 020: Protocol Buffer Client Encoding
+
+## Changelog
+
+- 2020 March 06: Initial Draft
+
+## Status
+
+Proposed
+
+## Context
+
+This ADR is a continuation of the motivation, design, and context established in
+[ADR 019](./adr-019-protobuf-state-encoding.md), namely, we aim to design the
+Protocol Buffer migration path for the client-side of the Cosmos SDK.
+
+Specifically, the client-side migration path primarily includes tx generation and
+signing, message construction and routing, in addition to CLI & REST handlers and
+business logic (i.e. queriers).
+
+With this in mind, we will tackle the migration path via two main areas, txs and
+querying.
+
+## Decision
+
+### Transactions
+
+Since the messages that an application is known and allowed to handle are specific
+to the application itself, so must the transactions be specific to the application
+itself. Similar to how we described in [ADR 019](./adr-019-protobuf-state-encoding.md),
+the concrete types will be defined at the application level via Protobuf `oneof`.
+
+The application will define a single canonical `Message` Protobuf message
+with a single `oneof` that implements the SDK's `Msg` interface.
+
+Example:
+
+```protobuf
+// app/codec/codec.proto
+
+message Message {
+  option (cosmos_proto.interface_type) = "github.com/cosmos/cosmos-sdk/types.Msg";
+
+  oneof sum {
+    bank.MsgSend = 1;
```

Comment:
I think this is where @aaronc's suggestion around module level `oneof` could come in handy by allowing devs to just register the modules here. Would like to hear him chime in here.

Reply:
Yes, option B from the [doc](https://docs.google.com/document/d/1eEgYgvgZqLE45vETjhwIw4VOqK-5hwQtZtjVbiXnIGc/edit#). A few things:

Option B can be done in parallel to this and the developer would have the option of either approach. But,

1. This requires reflection
2. Custom logic to be written by @aaronc or someone to do all the magic.

I don't want to hold up the ADR and implementation for that to be completed. So hence going with option A here. Option B can be done in parallel and devs would have the choice.

Reply:
I'll add a section on this in `Future Improvements`.

Reply:
That's fine.